extensions/vscode/src/entrypointTracker.ts

```diff
@@ -0,0 +1,214 @@
+   */
+  async activate(options?: { forceUpdate?: boolean }) {
+    // change based on if entrypoint
+    if (options?.forceUpdate) {
+      await this.update();
+    } else if (this.requiresUpdate) {
+      await this.update();
+    }
+
+    commands.executeCommand(
+      "setContext",
+      Contexts.ActiveFileEntrypoint,
+      this.isEntrypoint,
+    );
+  }
+
+  /**
+   * Updates the entrypoint next time the document is activated.
+   */
+  updateNextActivate() {
+    this.requiresUpdate = true;
+  }
+
+  /**
+   * Updates whether or not the document is an entrypoint file.
+   */
+  private async update() {
+    this.requiresUpdate = false;
+    this.isEntrypoint = await isDocumentEntrypoint(this.document);
+  }
+}
+
+/**
+ * Tracks active documents and assists in determining extension context.
+ */
+export class DocumentTracker implements Disposable {
+  private disposable: Disposable;
+
+  private readonly documents = new Map<
+    TextDocument,
+    TrackedEntrypointDocument
+  >();
+
+  constructor() {
+    this.disposable = Disposable.from(
+      window.onDidChangeActiveTextEditor(this.onActiveTextEditorChanged, this),
+      workspace.onDidCloseTextDocument(this.onTextDocumentClosed, this),
+      workspace.onDidSaveTextDocument(this.onTextDocumentSaved, this),
+    );
+
+    // activate the initial active file
+    this.onActiveTextEditorChanged(window.activeTextEditor);
+  }
+
+  dispose() {
+    this.disposable.dispose();
+    this.documents.clear();
+  }
+
+  /**
+   * Starts tracking a document
+   * @param document The document to track
+   * @returns The TrackedEntrypointDocument created
+   */
+  async addDocument(document: TextDocument) {
+    const entrypoint = await TrackedEntrypointDocument.create(document);
+    this.documents.set(document, entrypoint);
+    return entrypoint;
+  }
+
+  /**
+   * Stops tracking a document
+   * @param document The document to stop tracking
+   */
+  removeDocument(document: TextDocument) {
+    this.documents.delete(document);
+  }
+
+  /**
+   * Listener function for changes to the active text editor.
+   *
+   * Adds new documents to the tracker, and activates the associated
+   * TrackedEntrypointDocument
+   * @param editor The active text editor
+   */
+  async onActiveTextEditorChanged(editor: TextEditor | undefined) {
+    if (editor === undefined) {
+      commands.executeCommand(
+        "setContext",
+        Contexts.ActiveFileEntrypoint,
+        undefined,
+      );
+      return;
+    }
+
+    let tracked = this.documents.get(editor.document);
+
+    if (tracked === undefined) {
+      tracked = await this.addDocument(editor.document);
+    }
+
+    tracked.activate();
+  }
+
+  /**
+   * Listener function for the closing of a text document.
+   * Stops the document from being tracked.
+   *
+   * @param document The closed document
+   */
+  onTextDocumentClosed(document: TextDocument) {
+    this.removeDocument(document);
+  }
```

Comment:
Documents are removed when closed since another editor could potentially change the document. This way when a document is re-opened we recheck just in case.